From linux-usb-owner@vger.kernel.org  Wed Sep  1 12:48:46 2010
From: Praveena Nadahally <praveen.nadahally@stericsson.com>
To: <linux-usb@vger.kernel.org>
Cc: <STEricsson_nomadik_linux@list.st.com>,
	<linus.walleij@stericsson.com>, <dbrownell@users.sourceforge.net>,
	Parirajan Muthalagu <parirajan.muthalagu@stericsson.com>,
	Praveena Nadahally <praveen.nadahally@stericsson.com>
Subject: USB Gadget: Verify VBUS current before setting the device self-powered bit
Date: Wed, 25 Aug 2010 16:33:26 +0530
Message-ID: <1282734206-29760-1-git-send-email-praveen.nadahally@stericsson.com>

From: Parirajan Muthalagu <parirajan.muthalagu@stericsson.com>


Acked-by: David Brownell <dbrownell@users.sourceforge.net>
Acked-by: Linus Walleij <linus.walleij@stericsson.com>
Signed-off-by: Praveena Nadahally <praveen.nadahally@stericsson.com>
Signed-off-by: Parirajan Muthalagu <parirajan.muthalagu@stericsson.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/composite.c |    8 +++++++-
 include/linux/usb/ch9.h        |   10 ++++++++++
 2 files changed, 17 insertions(+), 1 deletion(-)

--- a/drivers/usb/gadget/composite.c
+++ b/drivers/usb/gadget/composite.c
@@ -1074,7 +1074,13 @@ static int composite_bind(struct usb_gad
 	cdev->bufsiz = USB_BUFSIZ;
 	cdev->driver = composite;
 
-	usb_gadget_set_selfpowered(gadget);
+	/*
+	 * As per USB compliance update, a device that is actively drawing
+	 * more than 100mA from USB must report itself as bus-powered in
+	 * the GetStatus(DEVICE) call.
+	 */
+	if (CONFIG_USB_GADGET_VBUS_DRAW <= USB_SELF_POWER_VBUS_MAX_DRAW)
+		usb_gadget_set_selfpowered(gadget);
 
 	/* interface and string IDs start at zero via kzalloc.
 	 * we force endpoints to start unassigned; few controller
--- a/include/linux/usb/ch9.h
+++ b/include/linux/usb/ch9.h
@@ -808,4 +808,14 @@ enum usb_device_state {
 	 */
 };
 
+/*-------------------------------------------------------------------------*/
+
+/*
+ * As per USB compliance update, a device that is actively drawing
+ * more than 100mA from USB must report itself as bus-powered in
+ * the GetStatus(DEVICE) call.
+ * http://compliance.usb.org/index.asp?UpdateFile=Electrical&Format=Standard#34
+ */
+#define USB_SELF_POWER_VBUS_MAX_DRAW		100
+
 #endif /* __LINUX_USB_CH9_H */
